crates/bevy_transform/src/systems.rs

```diff
@@ -89,100 +145,423 @@ pub fn propagate_transforms(
+    /// function will validate that the entity hierarchy does not contain cycles to prevent mutable
+    /// aliasing during propagation, but it is unable to verify that it isn't being used to mutably
+    /// alias the same entity.
+    ///
+    /// ## Panics
+    ///
+    /// Panics if the parent of a child node is not the same as the supplied `parent`. This
+    /// assertion ensures that the hierarchy is acyclic, which in turn ensures that if the caller is
+    /// following the supplied safety rules, multi-threaded propagation is sound.
+    #[inline]
+    #[expect(unsafe_code, reason = "Mutating disjoint entities in parallel")]
+    unsafe fn propagate_descendants_unchecked(
+        parent: Entity,
+        p_global_transform: Mut<GlobalTransform>,
+        p_children: &Children,
+        nodes: &NodeQuery,
+        outbox: &mut Vec<Entity>,
+        queue: &WorkQueue,
+        max_depth: usize,
+    ) {
+        // Create mutable copies of the input variables, used for iterative depth-first traversal.
+        let (mut parent, mut p_global_transform, mut p_children) =
+            (parent, p_global_transform, p_children);
+
+        // See the optimization note at the end to understand why this loop is here.
+        for depth in 1..=max_depth {
+            // Safety: traversing the entity tree from the roots, we assert that the childof and
+            // children pointers match in both directions (see assert below) to ensure the hierarchy
+            // does not have any cycles. Because the hierarchy does not have cycles, we know we are
+            // visiting disjoint entities in parallel, which is safe.
+            #[expect(unsafe_code, reason = "Mutating disjoint entities in parallel")]
+            let children_iter = unsafe {
+                // Performance note: iter_many tests every child to see if it meets the query. For
+                // leaf nodes, this unfortunately means we have the pay the price of checking every
+                // child, even if it is a leaf node and is skipped.
+                //
+                // To ensure this is still the fastest design, I tried removing the second pass
+                // (`compute_transform_leaves`) and instead simply doing that here. However, that
+                // proved to be much slower than two pass for a few reasons:
+                // - it's less cache friendly and is outright slower than the tight loop in the
+                //   second pass
+                // - it prevents parallelism, as all children must be iterated in series
+                //
+                // The only way I can see to make this faster when there are many leaf nodes is to
+                // speed up archetype checking to make the iterator skip leaf entities more quickly,
+                // or encoding the hierarchy level as a component. That, or use some kind of change
+                // detection to mark dirty subtrees when the transform is mutated.
```

Comment:
Yeah, I think we'll want to do the latter eventually. It sounds like what you have here is so fast that it tends to remove transformation propagation from the profile though.